### ISSUE TITLE
ROX(ReadOnlyMany) PVC for RBD seems to be wrong

### ISSUE THREAD

**Issue by lichaonw** [ceph/ceph-csi#1292]
# Describe the bug #
When i using ceph-csi image with lastest version, i want to use ROX PVC feature for rbd, but something is wrong.

# Environment details #

- Image/version of Ceph CSI driver : v3.0.0
- Helm chart version : none
- Kernel version : Linux version 3.10.0-957.el7.x86_64
- Mounter used for mounting PVC (for cephfs its `fuse` or `kernel`. for rbd its
  `krbd` or `rbd-nbd`) : krbd
- Kubernetes cluster version : 1.16.6
- Ceph cluster version : nautilus 14.29

# Steps to reproduce #

1. Deploy a storageclass
```yaml
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
   name: csi-rbd-sc-test
provisioner: rbd.csi.ceph.com
parameters:
   clusterID: 1fbfb12a-40cf-48c6-b6be-50cf007c6fe1
   imageFeatures: layering
   pool: kubernetes
   csi.storage.k8s.io/provisioner-secret-name: csi-rbd-secret
   csi.storage.k8s.io/provisioner-secret-namespace: default
   csi.storage.k8s.io/node-stage-secret-name: csi-rbd-secret
   csi.storage.k8s.io/node-stage-secret-namespace: default
   fsType: ext4
reclaimPolicy: Delete
mountOptions:
   - discard
   - ro
```
2. Deploy a PVC 
```yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: rbd-pvc-rom2-test
spec:
  accessModes:
    - ReadOnlyMany
  volumeMode: Block
  resources:
    requests:
      storage: 1Gi
  storageClassName: csi-rbd-sc-test
```
3. Deploy a pod with pvc
```yaml
---
apiVersion: v1
kind: Pod
metadata:
  name: csi-rbd-demo-pod4
spec:
  containers:
  - name: web-server
    image: 10.30.66.13:5000/busybox:latest
    command: [ "/bin/sh", "-ce", "tail -f /dev/null" ]
    volumeDevices:
      - name: mypvc
        devicePath: /dev/xvda
  volumes:
  - name: mypvc
    persistentVolumeClaim:
      claimName: rbd-pvc-rom2-test
```
4.  Deploy a pod with privileged
```yaml
---
apiVersion: v1
kind: Pod
metadata:
  name: csi-rbd-demo-pod4-1
spec:
  containers:
  - name: web-server
    image: 10.30.66.13:5000/busybox:latest
    command: [ "/bin/sh", "-ce", "tail -f /dev/null" ]
    securityContext:
      privileged: true
    volumeDevices:
      - name: mypvc
        devicePath: /dev/xvda
  volumes:
  - name: mypvc
    persistentVolumeClaim:
      claimName: rbd-pvc-rom2-test
```
# Actual results #

1. In pod without privileged, when i want to mount device, it show:
```
/ # mkdir /data
/ # mkfs.ext2 /dev/xvda 
Filesystem label=
OS type: Linux
Block size=4096 (log=2)
Fragment size=4096 (log=2)
65536 inodes, 262144 blocks
13107 blocks (5%) reserved for the super user
First data block=0
Maximum filesystem blocks=4194304
8 block groups
32768 blocks per group, 32768 fragments per group
8192 inodes per group
Superblock backups stored on blocks:
	32768, 98304, 163840, 229376
/ # mount  /dev/xvda /data
mount: permission denied (are you root?)
```
It show no root permission to mount, so i deploy anthoer pod with privileged, but it show
```
/ # ls /dev | grep xvda
```
it did not mount rbd volume  without any error message.

**Comment by Madhu-1** [comment]
Read-only is supported for only clone. Why someone will create read-only empty pvc? Is there any usecase?

**Comment by Madhu-1** [comment]
> Read-only is supported for the only clone. Why someone will create read-only empty PVC? Is there any use-case?

currently, it's not disabled it would be best to disable it. @dillaman any thoughts?

**Comment by Madhu-1** [comment]
```
apiVersion: v1
kind: Pod
metadata:
  name: my-pod
spec:
  containers:
    - name: my-container
      securityContext:
            privileged: true
            capabilities:
              add: ["SYS_ADMIN", "SYS_RAWIO"]
      image: busybox
      command: [ "/bin/sh", "-ce", "tail -f /dev/null" ]
      volumeDevices:
        - devicePath: /dev/block
          name: my-volume
      imagePullPolicy: IfNotPresent
  volumes:
    - name: my-volume
      persistentVolumeClaim:
        claimName: rbd-pvc
```
you need to set `SYS_RAWIO`  see https://kubernetes.io/blog/2019/03/07/raw-block-volume-support-to-beta/

**Comment by lichaonw** [comment]
> capabilities:
>               add: ["SYS_ADMIN", "SYS_RAWIO"]

@Madhu-1 it seems doesn't work

**Comment by Madhu-1** [comment]
Does the above pod works with readwritemany ?

**Comment by lichaonw** [comment]
> Does the above pod works with readwritemany ?

no, it work in readonlymany mode

**Comment by lichaonw** [comment]
@Madhu-1 this is  run message about my pod :

```yaml
Name:         csi-rbd-demo-pod4-1
Namespace:    default
Priority:     0
Node:         nm-bigdata-030033070.ctc.local/10.30.33.70
Start Time:   Thu, 30 Jul 2020 13:02:52 +0800
Labels:       <none>
Annotations:  kubectl.kubernetes.io/last-applied-configuration:
                {"apiVersion":"v1","kind":"Pod","metadata":{"annotations":{},"name":"csi-rbd-demo-pod4-1","namespace":"default"},"spec":{"containers":[{"c...
Status:       Running
IP:           10.244.3.232
IPs:
  IP:  10.244.3.232
Containers:
  web-server:
    Container ID:  docker://0cf2499fb90a74d8f13200a99b16e730f0fb7906c19ac8a937d0b1979266c557
    Image:         10.30.66.13:5000/busybox:latest
    Image ID:      docker-pullable://10.30.66.13:5000/busybox@sha256:edafc0a0fb057813850d1ba44014914ca02d671ae247107ca70c94db686e7de6
    Port:          <none>
    Host Port:     <none>
    Command:
      /bin/sh
      -ce
      tail -f /dev/null
    State:          Running
      Started:      Thu, 30 Jul 2020 13:02:54 +0800
    Ready:          True
    Restart Count:  0
    Environment:    <none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-zmhx5 (ro)
    Devices:
      /dev/xvda from mypvc
Conditions:
  Type              Status
  Initialized       True 
  Ready             True 
  ContainersReady   True 
  PodScheduled      True 
Volumes:
  mypvc:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  rbd-pvc-rom2-test
    ReadOnly:   false
  default-token-zmhx5:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  default-token-zmhx5
    Optional:    false
QoS Class:       BestEffort
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:          <none>
```

---
In my opinion, something wrong with 'ReadOnly:   false' in stage of mount volume

**Comment by Madhu-1** [comment]
@Yuggupta27 PTAL

**Comment by houz42** [comment]
> Read-only is supported for only clone. Why someone will create read-only empty pvc? Is there any usecase?

Sorry to reply to a closed issue. Cause kubernetes allows assigning more than one access modes to one PVC (RWO and ROX for example), I may want to create a PVC using ceph rbd like this, and I wish I could start a "data preparing" pod writting data to the PVC first, and stop the first pod and then start more than one pods mount the PVC read from the same PVC. I know I can achieve this workflow by cloning new pvcs based on snapshots, but if it could be done without snapshots and many other read-only pvcs cloned from the snapshots, the workflow may run faster and keep the k8s and ceph clean.   

So in short, my usecase if not "create read-only empty pvc", but create a pvc can be mounted read write and read only in different time. And as far as I know, this workflow is not forbidded by Ceph itself, right?

By the way, as digging into the csi implementation trying to workout this workflow, I found it's hard to be done correctly. I can not tell whether the pod wants to mount the pvc read only in node staging, if remove all inuse check, there will certainly be more mistakes.
